posthog/temporal/data_imports/pipelines/source/config.py

```diff
@@ -0,0 +1,299 @@
+                elif nested_prefix:
+                    # Prefer the class-level prefix if set
+                    nested_prefixes = (*top_level_prefixes, nested_prefix)
+                else:
+                    default_nested_prefix = _get_default_prefix_for_class(field_type)
+                    nested_prefixes = (*top_level_prefixes, default_nested_prefix)
+
+                value = to_config(field_type, d, nested_prefixes)
+
+        else:
+            key = _get_key(field, top_level_prefixes)
+
+            try:
+                value = d[key]
+            except KeyError:
+                continue
+
+        if field_metadata and field_metadata.converter:
+            convert = field_metadata.converter
+        else:
+            convert = _noop_convert
+
+        inputs[field.name] = convert(value)
+
+    return config_cls(**inputs)
+
+
+def _lookup_type(type_to_resolve: str, locals: dict[str, typing.Any], globals: dict[str, typing.Any]) -> type:
+    """Lookup a type in provided locals and globals.
+
+    Used to resolve any type hints that are strings.
+    """
+    try:
+        return locals[type_to_resolve]
+    except KeyError:
+        try:
+            return globals[type_to_resolve]
+        except KeyError:
+            raise TypeError(f"Unknown type: '{type_to_resolve}'")
+
+
+def _get_key(field: dataclasses.Field[typing.Any], prefixes: tuple[str, ...]) -> str:
+    """Get the key used to lookup a field."""
+    try:
+        metadata = field.metadata[META_KEY]
+    except KeyError:
+        return "_".join((*prefixes, field.name))
+
+    name = field.name
+    if metadata.lookup_name is not None:
+        name = metadata.lookup_name
+
+    prefixes = prefixes
+    if metadata.prefix is not None:
+        prefixes = metadata.prefix
+
+    return "_".join((*prefixes, name))
+
+
+def _get_default_prefix_for_class(cls: type) -> str:
+    """Get a default prefix for given class based on name.
+
+    This function attempts to extract words from a class name to form a string
+    with words separated by '_'.
+
+    In this context, word has multiple exclusive definitions:
+    * An uppercase character followed by a sequence of lowercase characters.
+    * A sequence of uppercase characters.
+
+    These are only heuristics though and there are cases in which we cannot
+    separate words, like if a class name is composed of multiple continuous
+    uppercase characters, e.g. "AWSKMSKey" would result in "awskms_key" instead
+    of perhaps the expected "aws_kms_key".
+
+    Examples:
+        >>> class SSHTunnel: ...
+        >>> _get_default_prefix_for_class(SSHTunnel)
+        'ssh_tunnel'
+        >>> class Test: ...
+        >>> _get_default_prefix_for_class(Test)
+        'test'
+        >>> class MyClass: ...
+        >>> _get_default_prefix_for_class(MyClass)
+        'my_class'
+        >>> class AWSKMSKey: ...
+        >>> _get_default_prefix_for_class(AWSKMSKey)
+        'awskms_key'
```

Comment:
All of these doctests pass, and can be run with:

```
DEBUG=1 pytest --doctest-modules 'posthog/temporal/data_imports/pipelines/source/config.py' -vv
```

Reply:
We also have unit tests for the high level api, but I at least wanted some examples for the low level stuff.